### ISSUE TITLE
Out of memory processing large FOV data

### ISSUE THREAD

**Issue by Michael-shannon** [MouseLand/cellpose#178]
I am attempting to segment large 3D montages of thousands of cells. Each z stack is 696 mb, and there are 219 of them. Image size is 2880 x 2880 px. 43 planes per z stack.

I received an OOM error on a decent machine with 64 gb RAM available. First, I tried to segment all 219 of the z stacks using the command line. This is where the first error message occurred, before any prediction could begin at the data loading stage:

`(cellpose2) C:\Users\3i>python -m cellpose --dir E:/mydata/ --pretrained_model nuclei --diameter 24 --save_png --use_gpu --do_3D --save_tif --batch_size 8
matplotlib not installed
** CUDA version installed and working. **
>>>> using GPU
Traceback (most recent call last):
  File "C:\Users\3i\.conda\envs\cellpose2\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\3i\.conda\envs\cellpose2\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "e:\cellposelatest\cellpose-master\cellpose\__main__.py", line 265, in <module>
    main()
  File "e:\cellposelatest\cellpose-master\cellpose\__main__.py", line 132, in main
    images = [io.imread(image_names[n]) for n in range(nimg)]
  File "e:\cellposelatest\cellpose-master\cellpose\__main__.py", line 132, in <listcomp>
    images = [io.imread(image_names[n]) for n in range(nimg)]
  File "e:\cellposelatest\cellpose-master\cellpose\io.py", line 40, in imread
    img = tifffile.imread(filename)
  File "C:\Users\3i\.conda\envs\cellpose2\lib\site-packages\tifffile\tifffile.py", line 715, in imread
    return tif.asarray(**kwargs)
  File "C:\Users\3i\.conda\envs\cellpose2\lib\site-packages\tifffile\tifffile.py", line 2944, in asarray
    typecode, product(series.shape), out=out
  File "C:\Users\3i\.conda\envs\cellpose2\lib\site-packages\tifffile\tifffile.py", line 8349, in read_array
    result = numpy.empty(count, dtype) if out is None else out
MemoryError`

Then, I tried cutting the number of available files down to only 10 files (each a z stack of 680 mb). This time, the above memory error was avoided, but another memory error occurred during processing of the first image (I think during the 'running zy' stage.

Finally, I tried running cellpose via command line on a single z stack. This seemed to work, but then failed to save the masks as png, and the seg.npy file is unreadable in the cellpose GUI. From that, I received the following error message:

`flows computed 7939.92s
dynamics computed 8010.48s
masks computed 8409.50s
estimated masks for 1 image(s) in 8988.74 sec
>>>> TOTAL TIME 8988.74 sec
>>>> saving results
Traceback (most recent call last):
  File "C:\Users\3i\.conda\envs\cellpose2\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\3i\.conda\envs\cellpose2\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "e:\cellposelatest\cellpose-master\cellpose\__main__.py", line 265, in <module>
    main()
  File "e:\cellposelatest\cellpose-master\cellpose\__main__.py", line 178, in main
    io.masks_flows_to_seg(images, masks, flows, diams, image_names, channels)
  File "e:\cellposelatest\cellpose-master\cellpose\io.py", line 187, in masks_flows_to_seg
    masks_flows_to_seg(image, mask, flow, diam, file_name, channels_img)
  File "e:\cellposelatest\cellpose-master\cellpose\io.py", line 219, in masks_flows_to_seg
    'est_diam': diams})
  File "C:\Users\3i\.conda\envs\cellpose2\lib\site-packages\numpy\lib\npyio.py", line 536, in save
    pickle_kwargs=pickle_kwargs)
  File "C:\Users\3i\.conda\envs\cellpose2\lib\site-packages\numpy\lib\format.py", line 633, in write_array
    pickle.dump(array, fp, protocol=2, **pickle_kwargs)
OverflowError: cannot serialize a string larger than 4GiB`

I am using the MXNET version of cellpose, but plan to update to the Pytorch version today and try it there.

It seems to be that possibly, it might be better behaviour for cellpose to treat each z stack separately, performing memory dumps in between. Is such a thing possible?

Most appreciatively,

Michael

**Comment by papalagirauscher** [comment]
Hey!

The problem with this is numpy not being able to process strings of this size. you need to change the numpy working mode. I ran into the same issue and was able to solve the problem. Google will help you further, sadly I don't remember it 100%.

**Comment by Michael-shannon** [comment]
Hey :) 

Thanks, that led to some digging. I can't find anything about changing the numpy mode at the moment.

I can find some stuff related to 'pickle', and that things might work better if I was able to change the pickle.dump protocol to 4 somewhere, instead of 2. I can find it myself given enough time, but as this is a fix that may help the community I will wait on that.

Others say that 4GiB is a hard coded limit, and so this would either require choice memory dumps or splitting files up into manageable chunks. I thought this might be already done by 'tiling' but perhaps that splits up files for normalizing the image size rather than for memory reasons?

Papalgirauscher - 

Do you think I am having the same issue between the two error messages?

(Update, I got the PyTorch version up and running, and I see the same problem).

M

**Comment by Michael-shannon** [comment]
Update: perhaps you are describing mmap_mode in numpy

https://stackoverflow.com/questions/42727412/efficient-way-to-partially-read-large-numpy-file

This sounds like a potential option?

**Comment by carsen-stringer** [comment]
The first issue you're having I think has been resolved in the latest version of cellpose that loops in the CLI? there is no longer a line in `__main__.py` that tries to load all the files (`images = [io.imread(image_names[n]) for n in range(nimg)]`). instead the loop loads a single image at a time. Can you try upgrading if you haven't already?

If you are having issues saving NPY outputs you can turn that output off with the `--no_npy` flag in the command line interface. but if you're running out of memory due to the size of the tiff we don't currently have a workaround for that -- we recommend splitting the stack and stitching the masks afterwards.

**Comment by Michael-shannon** [comment]
Thanks Carsen. 

Yeah, the above was all from the new version that loops in the CLI.

The --no_npy flag is a good tip, but didn't solve the problem I was having and I still run out of memory (VRAM). I will investigate splitting and stitching, yes. I guess whatever I use to do that, the stitching can be computed based on the masks, and the same operation can be applied to the raw image data (rather than computing two separate stitchings, one for raw, one for masks).

**Comment by papalagirauscher** [comment]
Hello!
I tracked down how to create larger npy files. You need to change the protocol in numpy to 4. 

python3.7/site-packages/numpy/lib/format.py

Change line 664 
pickle.dump(array, fp, protocol=3 ...

to
pickle.dump(array, fp, protocol=4 ...

Best wishes,
Jan

**Comment by Michael-shannon** [comment]
Thanks! That seems to have worked. I can now produce .npy files from my largest images so far. The individual (z stack) images are about 700 mb each, whereas the npy file for each is 13 gb. I agree with Carsen and Marius that perhaps using npy arrays in this way for saving the data is suboptimal. I think they are checking whether there is a better format that can be used by all, for downstream processing. For now, just the tif masks are pretty good.

**Comment by carsen-stringer** [comment]
closing this issue, I agree that the npy format is not great, for now we'll keep it as the default, but I think soon we will only save the mask images
